From 15b2671e28989d6ab875facb49e39904283e6d2f Mon Sep 17 00:00:00 2001
From: Simon Xue <xxm@rock-chips.com>
Date: Wed, 11 Mar 2020 15:55:14 +0800
Subject: [PATCH] init: Add support directly parse cpio data decompressed by
 hardware decompressor

The working flow of new feature is:
->bootloader decompress ramdisk.gz to cpio data first and start kernel.
  ->decompressor driver notify initramfs that cpio data ready.
    ->initramfs continue to flush cpio data to rootfs.

Change-Id: I9cd5708fc93270ce77376c26d9da5a5c219996c1
Signed-off-by: Simon Xue <xxm@rock-chips.com>
Signed-off-by: Tao Huang <huangtao@rock-chips.com>
---
 include/linux/initramfs.h | 4 ++++
 init/initramfs.c          | 3 +++
 2 files changed, 7 insertions(+)

diff --git a/include/linux/initramfs.h b/include/linux/initramfs.h
index fc7da63b125b..0e9b15a00975 100644
--- a/include/linux/initramfs.h
+++ b/include/linux/initramfs.h
@@ -29,4 +29,8 @@ int __init default_rootfs(void);
 
 #endif
 
+#if defined(CONFIG_ROCKCHIP_HW_DECOMPRESS)
+void __init wait_initrd_hw_decom_done(void);
+#endif
+
 #endif /* _LINUX_INITRAMFS_H */
diff --git a/init/initramfs.c b/init/initramfs.c
index eb202c0eacc6..e36a911e4dd2 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -458,6 +458,9 @@ static char * __init unpack_to_rootfs(char *buf, unsigned long len)
 	state = Start;
 	this_header = 0;
 	message = NULL;
+#if defined(CONFIG_ROCKCHIP_THUNDER_BOOT) && defined(CONFIG_ROCKCHIP_HW_DECOMPRESS) && defined(CONFIG_INITRD_ASYNC)
+	wait_initrd_hw_decom_done();
+#endif
 	while (!message && len) {
 		loff_t saved_offset = this_header;
 		if (*buf == '0' && !(this_header & 3)) {
-- 
2.35.3

